src/dune/workspace.ml

```diff
@@ -165,29 +165,30 @@ module Context = struct
       let open Dyn.Encoder in
       record
         [ ("base", Common.to_dyn base)
-        ; ("switch", Context_name.to_dyn switch)
+        ; ("switch", string switch)
         ; ("root", option string root)
         ; ("merlin", bool merlin)
         ]
 
     let equal { base; switch; root; merlin } t =
       Common.equal base t.base
-      && Context_name.equal switch t.switch
+      && String.equal switch t.switch
       && Option.equal String.equal root t.root
       && Bool.equal merlin t.merlin
 
     let t ~profile ~x =
-      let+ switch = field "switch" Context_name.decode
+      let+ loc_switch, switch = field "switch" (located string)
       and+ name = field_o "name" Context_name.decode
       and+ root = field_o "root" string
       and+ merlin = field_b "merlin"
       and+ base = Common.t ~profile in
-      let default =
-        (* TODO this needs proper error handling with locations *)
-        let name = Context_name.to_string switch ^ Common.fdo_suffix base in
-        Context_name.parse_string_exn (Loc.none, name)
+      let name =
+        match name with
+        | Some s -> s
+        | None ->
+          let name = Filename.basename switch ^ Common.fdo_suffix base in
```

Comment:
This is slightly hacky because::

* The generated name is not necessarily a valid context name.
* The generated name might collide with other context names in the workspace file.

I'm open to suggestions on how to better address this.

Reply:
What about we don't do anything clever and simply ask the user to specify a name explicitly if the switch name is not a valid context name?

Reply:
That would require setting a name for all local switches. Seems a bit tedious.

Reply:
Can there be more than 1 local switch?

Reply:
Yes. However, it's not necessary for the name conflict. The following scenario is also possible:

```
(context (opam (switch /foo/bar))) ;; local switch
(context (opam (switch bar))) ;; global switch named bar
```

dune will infer the same context name and there will be an error later.

Reply:
why is it tedious to set a name for each local switch?

Reply:
Yeah, I guess it's not so bad. I changed the code to always require the `name` to be manually specified if it's not a valid context.